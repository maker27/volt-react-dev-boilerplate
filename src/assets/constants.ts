export const appName = 'Invoice App';

export const routes = {
    invoices: { url: '/', title: 'Invoices List' },
    invoices_create: { url: '/invoices/new', title: 'Create Invoice' },
    invoices_edit: { url: '/invoices/:id/edit', title: 'Edit Invoice' },
    products: { url: '/products', title: 'Products List' },
    customers: { url: '/customers', title: 'Customers List' }
};

export const apiBaseUrl = 'http://localhost:8000/api/';
