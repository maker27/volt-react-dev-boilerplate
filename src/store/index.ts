import { configureStore } from '@reduxjs/toolkit';
import invoicesReducer from './invoiceSlice';
import productReducer from './productSlice';
import customerReducer from './customerSlice';
import statusReducer from './statusSlice';

export const store = configureStore({
    reducer: {
        invoices: invoicesReducer,
        products: productReducer,
        customers: customerReducer,
        status: statusReducer
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
