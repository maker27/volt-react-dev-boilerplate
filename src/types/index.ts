import React from 'react';
import { IInvoice, TInvoice } from './invoice';
import { IProduct, TProduct } from './product';
import { ICustomer, TCustomer } from './customer';
import { IInvoiceItem, TInvoiceItem } from './invoice-item';

export type { IInvoice, IProduct, ICustomer, IInvoiceItem, TCustomer, TProduct, TInvoice, TInvoiceItem };

export type TOnChangeEvent = React.ChangeEvent<HTMLInputElement>;

export type TObjectOf<T> = { [key: string]: T };

export type TFormDataSetter<T> = (value: T) => void;
