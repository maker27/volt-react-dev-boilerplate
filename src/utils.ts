import { IInvoiceItem, IProduct, TObjectOf } from './types';

export function arrayToObject<T extends { id: number }>(list: T[], prop: keyof T = 'id'): TObjectOf<T> {
    return list.reduce((s, v) => ({ ...s, [String(v[prop])]: v }), {});
}

export function decimalNumber(value: number, decimalPlaces = 2): number {
    return Number(Math.round(parseFloat(value + 'e' + decimalPlaces)) + 'e-' + decimalPlaces);
}

export const getInvoiceTotal = (
    invoiceItems: TObjectOf<IInvoiceItem>,
    products: TObjectOf<IProduct>,
    discount = 0
): number => {
    const sum = Object.values(invoiceItems).reduce((sum: number, { product_id, quantity }) => {
        const product = products[product_id];
        return product ? sum + product.price * quantity : sum;
    }, 0);
    return decimalNumber(((100 - discount) / 100) * sum);
};
