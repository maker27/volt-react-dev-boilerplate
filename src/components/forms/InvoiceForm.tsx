import React, { FC, MouseEventHandler, useEffect, useState } from 'react';
import SimpleTable from '../SimpleTable';
import { Button, Form, InputGroup } from 'react-bootstrap';
import LabeledInput from './LabeledInput';
import LabeledSelect from './LabeledSelect';
import {
    ICustomer,
    IInvoice,
    IInvoiceItem,
    IProduct,
    TInvoice,
    TInvoiceItem,
    TObjectOf,
    TOnChangeEvent
} from '../../types';
import { getInvoiceTotal } from '../../utils';

const toSelectOption = ({ id, name }: ICustomer | IProduct) => ({ value: id, label: name });

type TInvoiceFormProps = {
    title: string;
    customers: TObjectOf<ICustomer>;
    products: TObjectOf<IProduct>;
    invoice: IInvoice;
    invoiceItems: TObjectOf<IInvoiceItem>;
    invoiceId: number;
    setCurrentInvoice: (invoice?: IInvoice) => void;
    addInvoiceItem: (invoiceItem: TInvoiceItem) => void;
    editInvoiceItem: (invoiceItem: IInvoiceItem) => void;
    deleteInvoiceItem: (invoiceItem: IInvoiceItem) => void;
};

const DeleteButton: FC<{ onClick?: MouseEventHandler }> = ({ onClick }) => {
    return (
        <button type="button" className="btn btn-outline-danger" onClick={onClick}>
            <i className="bi bi-trash"> </i> Delete
        </button>
    );
};

const InvoiceForm: FC<TInvoiceFormProps> = ({
    title,
    customers,
    products,
    invoice,
    invoiceItems,
    invoiceId,
    setCurrentInvoice,
    addInvoiceItem,
    editInvoiceItem,
    deleteInvoiceItem
}) => {
    const [productId, setProductId] = useState<number>(0);
    const setInvoiceProp = (prop: keyof TInvoice) => ({ target }: TOnChangeEvent) => {
        if (prop === 'discount') {
            const number = +target.value;
            if (number < 0 || number > 100) return;
        }
        setCurrentInvoice({ ...invoice, [prop]: target.value });
    };
    const currentCustomer = customers[invoice.customer_id];
    const currentProduct = products[productId];

    useEffect(() => {
        const total = getInvoiceTotal(invoiceItems, products, invoice.discount);
        if (invoice.total !== total) {
            setCurrentInvoice({ ...invoice, total });
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [invoiceItems, invoice.discount]);

    return (
        <>
            <h2>{title}</h2>
            <Form className="w-50">
                <LabeledInput
                    className="w-50"
                    id="form-invoice-discount"
                    label="Discount (%)"
                    type="number"
                    value={invoice.discount.toString()}
                    onChange={setInvoiceProp('discount')}
                />
                <LabeledSelect
                    id="form-invoice-customer"
                    label="Customer"
                    value={currentCustomer ? toSelectOption(currentCustomer) : null}
                    options={Object.values(customers).map(toSelectOption)}
                    onChange={selectedOption => {
                        setCurrentInvoice({ ...invoice, customer_id: selectedOption?.value ?? 0 });
                    }}
                />
                <InputGroup className="mb-3">
                    <LabeledSelect
                        className="w-75"
                        id="form-invoice-product"
                        label="Add product"
                        value={currentProduct ? toSelectOption(currentProduct) : null}
                        options={Object.values(products).map(toSelectOption)}
                        onChange={selectedOption => {
                            setProductId(selectedOption?.value ?? 0);
                        }}
                    />
                    <Button
                        variant="outline-secondary"
                        className="button-invoice-add"
                        onClick={() => {
                            if (productId) {
                                const addedItem = Object.values(invoiceItems).find(
                                    ({ product_id }) => product_id === productId
                                );
                                if (addedItem) {
                                    editInvoiceItem({
                                        ...addedItem,
                                        quantity: addedItem.quantity + 1
                                    });
                                } else {
                                    addInvoiceItem({
                                        invoice_id: invoiceId,
                                        product_id: productId,
                                        quantity: 1
                                    });
                                }
                            }
                        }}>
                        Add
                    </Button>
                </InputGroup>
            </Form>
            <SimpleTable
                columns={['Name', 'Price', 'Qty', '']}
                rows={Object.values(invoiceItems).map(invoiceItem => {
                    const { id, product_id, quantity } = invoiceItem;
                    const product = products[product_id];
                    return [
                        product.name,
                        product.price,
                        <Form.Control
                            className="w-25"
                            key={'product' + id}
                            type="number"
                            step="1"
                            value={quantity}
                            onChange={({ target }) => {
                                editInvoiceItem({
                                    ...invoiceItem,
                                    quantity: +target.value
                                });
                            }}
                        />,
                        <DeleteButton
                            key={'delete' + id}
                            onClick={() => {
                                deleteInvoiceItem(invoiceItem);
                            }}
                        />
                    ];
                })}
            />
            <h4>Total: {invoice.total}</h4>
        </>
    );
};

export default InvoiceForm;
