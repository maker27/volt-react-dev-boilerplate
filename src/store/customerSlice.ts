import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ICustomer, TObjectOf } from '../types';
import { arrayToObject } from '../utils';

interface customerState {
    all: TObjectOf<ICustomer>;
    current: ICustomer;
}

const emptyCustomer = { id: 0, name: '', address: '', phone: '' };

const initialState: customerState = {
    all: {},
    current: emptyCustomer
};

export const customerSlice = createSlice({
    name: 'customer',
    initialState,
    reducers: {
        setCustomers: (state, action: PayloadAction<Array<ICustomer>>) => {
            state.all = arrayToObject<ICustomer>(action.payload);
        },
        addCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            state.all[customer.id] = customer;
        },
        editCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            state.all[customer.id] = customer;
        },
        deleteCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            delete state.all[customer.id];
        },
        setCurrentCustomer: (state, action: PayloadAction<ICustomer | undefined>) => {
            state.current = action.payload || emptyCustomer;
        }
    }
});

export const {
    setCustomers,
    addCustomer,
    editCustomer,
    deleteCustomer,
    setCurrentCustomer
} = customerSlice.actions;

export default customerSlice.reducer;
