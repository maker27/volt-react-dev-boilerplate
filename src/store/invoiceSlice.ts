import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IInvoice, IInvoiceItem, TObjectOf } from '../types';
import { arrayToObject } from '../utils';

interface IInvoiceState {
    all: TObjectOf<IInvoice>;
    current: IInvoice;
    items: TObjectOf<IInvoiceItem>;
}

const emptyInvoice = {
    id: 0,
    customer_id: 0,
    discount: 0,
    total: 0
};

const initialState: IInvoiceState = {
    all: {},
    current: emptyInvoice,
    items: {}
};

export const invoiceSlice = createSlice({
    name: 'invoice',
    initialState,
    reducers: {
        setInvoices: (state, action: PayloadAction<Array<IInvoice>>) => {
            state.all = arrayToObject<IInvoice>(action.payload);
        },
        addInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            state.all[invoice.id] = invoice;
            state.current = invoice;
        },
        editInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            state.all[invoice.id] = invoice;
            state.current = invoice;
        },
        deleteInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            delete state.all[invoice.id];
        },
        setCurrentInvoice: (state, action: PayloadAction<IInvoice | undefined>) => {
            state.current = action.payload || emptyInvoice;
        },
        setInvoiceItems: (state, action: PayloadAction<IInvoiceItem[]>) => {
            state.items = arrayToObject<IInvoiceItem>(action.payload, 'product_id');
        },
        addInvoiceItem: (state, action: PayloadAction<IInvoiceItem>) => {
            const invoiceItem = action.payload;
            state.items[invoiceItem.product_id] = invoiceItem;
        },
        deleteInvoiceItem: (state, action: PayloadAction<IInvoiceItem>) => {
            const invoiceItem = action.payload;
            delete state.items[invoiceItem.product_id];
        }
    }
});

export const {
    setInvoices,
    addInvoice,
    editInvoice,
    deleteInvoice,
    setCurrentInvoice,
    setInvoiceItems,
    addInvoiceItem,
    deleteInvoiceItem
} = invoiceSlice.actions;

export default invoiceSlice.reducer;
