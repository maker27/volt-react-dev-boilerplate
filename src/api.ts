import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import { apiBaseUrl } from './assets/constants';
import { store } from './store';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';
import {
    IInvoice,
    IProduct,
    ICustomer,
    TCustomer,
    TProduct,
    TInvoice,
    IInvoiceItem,
    TInvoiceItem
} from './types';
import {
    addInvoice,
    addInvoiceItem,
    deleteInvoice,
    deleteInvoiceItem,
    editInvoice,
    editInvoiceItem,
    setInvoiceItems,
    setInvoices
} from './store/invoiceSlice';
import { addProduct, deleteProduct, editProduct, setProducts } from './store/productSlice';
import { addCustomer, deleteCustomer, editCustomer, setCustomers } from './store/customerSlice';
import { StatusType, addStatus, setLoadAction } from './store/statusSlice';

const server = axios.create({
    baseURL: apiBaseUrl
});

type TRequestData = { [key: string]: string | number };

const isInitialRequest = ({ method, url = '' }: AxiosRequestConfig) =>
    method === 'get' && ['customers', 'products', 'invoices'].includes(url);

class API {
    constructor() {
        server.interceptors.request.use((req: AxiosRequestConfig) => {
            if (!isInitialRequest(req)) API.setLoader(true);
            return req;
        });
        server.interceptors.response.use((res: AxiosResponse) => {
            if (!isInitialRequest(res.config)) API.setLoader(false);
            return res;
        });
    }

    private static setLoader(load: boolean) {
        store.dispatch(setLoadAction(load));
        return load;
    }

    private static setError(e: Error) {
        API.setLoader(false);
        store.dispatch(addStatus({ id: Date.now(), type: StatusType.error, value: e.message }));
        console.error(e);
    }

    private request<T>(
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        serverRequest: <T = any, R = AxiosResponse<T>>(
            url: string,
            config?: AxiosRequestConfig
        ) => Promise<R>,
        params: [url: string, data?: TRequestData],
        action: ActionCreatorWithPayload<T>
    ) {
        return serverRequest(...params)
            .then(({ data }) => {
                store.dispatch(action(data));
                return data;
            })
            .catch(e => API.setError(e));
    }

    private getRequest<T>(url: string, action: ActionCreatorWithPayload<T>) {
        return this.request(server.get, [url], action);
    }

    private postRequest<T>(url: string, data: TRequestData, action: ActionCreatorWithPayload<T>) {
        return this.request(server.post, [url, data], action);
    }

    private putRequest<T>(url: string, data: TRequestData, action: ActionCreatorWithPayload<T>) {
        return this.request(server.put, [url, data], action);
    }

    private deleteRequest<T>(url: string, action: ActionCreatorWithPayload<T>) {
        return this.request(server.delete, [url], action);
    }

    getInvoices() {
        return this.getRequest<IInvoice[]>('invoices', setInvoices);
    }

    getProducts() {
        return this.getRequest<IProduct[]>('products', setProducts);
    }

    getCustomers() {
        return this.getRequest<ICustomer[]>('customers', setCustomers);
    }

    getAllData() {
        return Promise.all([this.getInvoices(), this.getProducts(), this.getCustomers()]);
    }

    createCustomer(customerData: TCustomer) {
        this.postRequest<ICustomer>('customers', customerData, addCustomer);
    }

    editCustomer(customerId: number, customerData: TCustomer) {
        this.putRequest<ICustomer>(
            'customers/' + customerId,
            { ...customerData, customer_id: customerId },
            editCustomer
        );
    }

    deleteCustomer(customerId: number) {
        this.deleteRequest<ICustomer>('customers/' + customerId, deleteCustomer);
    }

    createProduct(productData: TProduct) {
        this.postRequest<IProduct>('products', productData, addProduct);
    }

    editProduct(productId: number, productData: TProduct) {
        this.putRequest<IProduct>(
            'products/' + productId,
            { ...productData, product_id: productId },
            editProduct
        );
    }

    deleteProduct(productId: number) {
        this.deleteRequest<IProduct>('products/' + productId, deleteProduct);
    }

    createInvoice(invoiceData: TInvoice) {
        return this.postRequest<IInvoice>('invoices', invoiceData, addInvoice);
    }

    editInvoice(invoiceId: number, invoiceData: Partial<TInvoice>) {
        this.putRequest<IInvoice>('invoices/' + invoiceId, invoiceData, editInvoice);
    }

    deleteInvoice(invoiceId: number) {
        this.deleteRequest<IInvoice>('invoices/' + invoiceId, deleteInvoice);
    }

    getInvoiceItems(invoiceId: number) {
        return this.getRequest<IInvoiceItem[]>('invoices/' + invoiceId + '/items', setInvoiceItems);
    }

    addInvoiceItem(invoiceItemData: TInvoiceItem) {
        return this.postRequest<IInvoiceItem>(
            'invoices/' + invoiceItemData.invoice_id + '/items',
            invoiceItemData,
            addInvoiceItem
        );
    }

    editInvoiceItem({ id, ...invoiceItemData }: IInvoiceItem) {
        this.putRequest<IInvoiceItem>(
            'invoices/' + invoiceItemData.invoice_id + '/items/' + id,
            { ...invoiceItemData },
            editInvoiceItem
        );
    }

    deleteInvoiceItem({ id, invoice_id }: IInvoiceItem) {
        this.deleteRequest<IInvoiceItem>('invoices/' + invoice_id + '/items/' + id, deleteInvoiceItem);
    }

    addNewInvoice(invoice: IInvoice, invoiceItems: TInvoiceItem[]) {
        API.setLoader(true);
        this.createInvoice(invoice).then(({ id }) => {
            Promise.all(
                Object.values(invoiceItems).map(invoiceItem =>
                    this.addInvoiceItem({ ...invoiceItem, invoice_id: id })
                )
            ).then(() => {
                API.setLoader(false);
            });
        });
    }
}

export default new API();
