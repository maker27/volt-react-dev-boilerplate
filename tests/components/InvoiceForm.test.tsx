import React from 'react';
import { render, fireEvent, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../src/store';
import { Route, MemoryRouter } from 'react-router-dom';
import InvoiceForm from '../../src/components/forms/InvoiceForm';
import { IInvoice, IInvoiceItem } from '../../src/types';
import { addInvoice, addInvoiceItem, setCurrentInvoice } from '../../src/store/invoiceSlice';
import { testCustomersList, testInvoiceItemsList, testInvoicesList, testProductsList } from '../data';
import { routes } from '../../src/assets/constants';
import { arrayToObject } from '../../src/utils';
import selectEvent from 'react-select-event';

const deleteInvoiceItem = jest.fn();

const renderForm = (): RenderResult => {
    const { invoices } = store.getState();
    const setInvoice = (newInvoice?: IInvoice) => store.dispatch(setCurrentInvoice(newInvoice));
    const setInvoiceItem = (newInvoiceItem: IInvoiceItem) => store.dispatch(addInvoiceItem(newInvoiceItem));

    return render(
        <Provider store={store}>
            <MemoryRouter initialEntries={[`/invoices/new`]}>
                <Route path="/invoices/new">
                    <InvoiceForm
                        title={routes.invoices_create.title}
                        customers={arrayToObject(testCustomersList)}
                        products={arrayToObject(testProductsList)}
                        invoice={invoices.current}
                        invoiceItems={arrayToObject(testInvoiceItemsList)}
                        invoiceId={0}
                        setCurrentInvoice={setInvoice}
                        addInvoiceItem={invoiceItem => setInvoiceItem({ ...invoiceItem, id: Date.now() })}
                        editInvoiceItem={setInvoiceItem}
                        deleteInvoiceItem={deleteInvoiceItem}
                    />
                </Route>
            </MemoryRouter>
        </Provider>
    );
};

const getCurrentInvoice = (): IInvoice => {
    const invoice = store.getState().invoices.current;
    expect(invoice).not.toBeUndefined();
    return invoice as IInvoice;
};

describe('InvoiceForm', () => {
    const addedInvoice = testInvoicesList[0];
    beforeAll(() => {
        store.dispatch(addInvoice(addedInvoice));
    });

    test('renders form', () => {
        const { asFragment } = renderForm();
        expect(asFragment()).toMatchSnapshot();

        expect(getCurrentInvoice()).toEqual(addedInvoice);
    });

    test('changes customer select value', async () => {
        const otherCustomer = testCustomersList[1];

        const { getByLabelText } = renderForm();
        const customerSelect = getByLabelText('Customer') as HTMLSelectElement;
        expect(customerSelect.value).toBe('');
        await selectEvent.select(customerSelect, otherCustomer.name);
        expect(getCurrentInvoice()).toEqual({ ...addedInvoice, customer_id: otherCustomer.id });
    });

    test('add new invoice items', async () => {
        const addedItem = testInvoiceItemsList[1];
        const addedItemName = testProductsList.find(({ id }) => id === addedItem.product_id)?.name ?? '';

        const { getByLabelText, getByText } = renderForm();
        const productSelect = getByLabelText('Add product');
        await selectEvent.select(productSelect, addedItemName);
        fireEvent.click(getByText('Add'));

        const invoiceItems = store.getState().invoices.items;
        expect(invoiceItems).toMatchInlineSnapshot(`
                  Object {
                    "${addedItem.product_id}": Object {
                      "id": ${addedItem.id},
                      "invoice_id": ${addedItem.invoice_id},
                      "product_id": ${addedItem.product_id},
                      "quantity": ${addedItem.quantity + 1},
                    },
                  }
                `);
    });
});
