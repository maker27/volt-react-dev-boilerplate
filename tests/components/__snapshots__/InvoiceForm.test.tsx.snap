// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceForm renders form 1`] = `
<DocumentFragment>
  <h2>
    Create Invoice
  </h2>
  <form
    class="w-50"
  >
    <div
      class="mb-3 w-50"
    >
      <label
        class="fw-bold form-label"
        for="form-invoice-discount"
      >
        Discount (%)
      </label>
      <input
        class="form-control"
        id="form-invoice-discount"
        maxlength="255"
        type="number"
        value="3"
      />
    </div>
    <div
      class="mb-3 "
    >
      <label
        class="fw-bold form-label"
        for="form-invoice-customer"
      >
        Customer
      </label>
      <div
        class=" css-2b097c-container"
      >
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class=" css-1uccc91-singleValue"
            >
              Test Customer
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="form-invoice-customer"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                />
              </svg>
            </div>
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="form-field-name"
          type="hidden"
          value="1"
        />
      </div>
    </div>
    <div
      class="mb-3 input-group"
    >
      <div
        class="mb-3 w-75"
      >
        <label
          class="fw-bold form-label"
          for="form-invoice-product"
        >
          Add product
        </label>
        <div
          class=" css-2b097c-container"
        >
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-yk16xz-control"
          >
            <div
              class=" css-g1d714-ValueContainer"
            >
              <div
                class=" css-1wa3eu0-placeholder"
              >
                Select...
              </div>
              <div
                class="css-b8ldur-Input"
              >
                <div
                  class=""
                  style="display: inline-block;"
                >
                  <input
                    aria-autocomplete="list"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    id="form-invoice-product"
                    spellcheck="false"
                    style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="form-field-name"
            type="hidden"
            value=""
          />
        </div>
      </div>
      <button
        class="button-invoice-add btn btn-outline-secondary"
        type="button"
      >
        Add
      </button>
    </div>
  </form>
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Price
        </th>
        <th>
          Qty
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="align-middle"
        >
          Product 101
        </td>
        <td
          class="align-middle"
        >
          23.39
        </td>
        <td
          class="align-middle"
        >
          <input
            class="w-25 form-control"
            step="1"
            type="number"
            value="2"
          />
        </td>
        <td
          class="align-middle"
        >
          <button
            class="btn btn-outline-danger"
            type="button"
          >
            <i
              class="bi bi-trash"
            >
               
            </i>
             Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          class="align-middle"
        >
          Product 102
        </td>
        <td
          class="align-middle"
        >
          17
        </td>
        <td
          class="align-middle"
        >
          <input
            class="w-25 form-control"
            step="1"
            type="number"
            value="1"
          />
        </td>
        <td
          class="align-middle"
        >
          <button
            class="btn btn-outline-danger"
            type="button"
          >
            <i
              class="bi bi-trash"
            >
               
            </i>
             Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <h4>
    Total: 61.87
  </h4>
</DocumentFragment>
`;
